File: e2e/java/src/test/java/dev/kreuzberg/e2e/CrawlTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import dev.kreuzberg.kreuzcrawl.Kreuzcrawl;

/** E2e tests for category: crawl. */
class CrawlTest {
    @Test
    void testContentBinarySkip() throws Exception {
        // Skips image and video content types gracefully
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/content_binary_skip";
        var result = Kreuzcrawl.scrape(engine, url);
        assertEquals(true, result.wasSkipped());
    }

    @Test
    void testContentPdfLinkSkip() throws Exception {
        // Encounters PDF link and skips or marks as document type
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/content_pdf_link_skip";
        var result = Kreuzcrawl.scrape(engine, url);
        assertEquals(true, result.wasSkipped());
    }

    @Test
    void testCrawlConcurrentDepth() throws Exception {
        // Concurrent crawl respects max_depth limit
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_concurrent_depth";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
        // skipped: field 'stayed_on_domain' not available on result type
    }

    @Test
    void testCrawlConcurrentLimit() throws Exception {
        // Respects max concurrent requests limit during crawl
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_concurrent_limit";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlConcurrentMaxPages() throws Exception {
        // Concurrent crawl respects max_pages budget
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_concurrent_max_pages";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlCustomHeaders() throws Exception {
        // Sends custom headers on all crawl requests
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_custom_headers";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlDepthOne() throws Exception {
        // Follows links one level deep from start page
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_depth_one";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
        // skipped: field 'stayed_on_domain' not available on result type
    }

    @Test
    void testCrawlDepthPriority() throws Exception {
        // Crawls in breadth-first order, processing depth-0 pages before depth-1
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_depth_priority";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlDepthTwo() throws Exception {
        // Crawls 3 levels deep (depth 0, 1, 2)
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_depth_two";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlDepthTwoChain() throws Exception {
        // Depth=2 crawl follows a chain of links across three levels
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_depth_two_chain";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlDoubleSlashNormalization() throws Exception {
        // Normalizes double slashes in URL paths (//page to /page)
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_double_slash_normalization";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'unique_urls.length' not available on result type
    }

    @Test
    void testCrawlEmptyPageNoLinks() throws Exception {
        // Crawl completes when child page has no outgoing links
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_empty_page_no_links";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlExcludePathPattern() throws Exception {
        // Skips URLs matching the exclude path pattern
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_exclude_path_pattern";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlExternalLinksIgnored() throws Exception {
        // External links are discovered but not followed when stay_on_domain is true
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_external_links_ignored";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
        // skipped: field 'stayed_on_domain' not available on result type
    }

    @Test
    void testCrawlFragmentStripping() throws Exception {
        // Strips #fragment from URLs for deduplication
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_fragment_stripping";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'unique_urls.length' not available on result type
    }

    @Test
    void testCrawlIncludePathPattern() throws Exception {
        // Only follows URLs matching the include path pattern
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_include_path_pattern";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlMaxDepthZero() throws Exception {
        // max_depth=0 crawls only the seed page with no link following
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_max_depth_zero";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlMaxPages() throws Exception {
        // Stops crawling at page budget limit
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_max_pages";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlMixedContentTypes() throws Exception {
        // Crawl handles links to non-HTML content types gracefully
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_mixed_content_types";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlMultipleRedirectsInTraversal() throws Exception {
        // Multiple linked pages with redirects are handled during crawl traversal
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_multiple_redirects_in_traversal";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlQueryParamDedup() throws Exception {
        // Deduplicates URLs with same query params in different order
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_query_param_dedup";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'unique_urls.length' not available on result type
    }

    @Test
    void testCrawlRedirectInTraversal() throws Exception {
        // Links that redirect are followed during crawl traversal
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_redirect_in_traversal";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlSelfLinkNoLoop() throws Exception {
        // Page linking to itself does not cause infinite crawl loop
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_self_link_no_loop";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlSinglePageNoLinks() throws Exception {
        // Crawling a page with no links returns only the seed page
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_single_page_no_links";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlStayOnDomain() throws Exception {
        // Does not follow external links when stay_on_domain is true
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_stay_on_domain";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
        // skipped: field 'stayed_on_domain' not available on result type
    }

    @Test
    void testCrawlSubdomainExclusion() throws Exception {
        // Stays on exact domain and skips subdomain links
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_subdomain_exclusion";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
        // skipped: field 'stayed_on_domain' not available on result type
    }

    @Test
    void testCrawlSubdomainInclusion() throws Exception {
        // Crawls subdomains when allow_subdomains is enabled
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_subdomain_inclusion";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

    @Test
    void testCrawlTrailingSlashDedup() throws Exception {
        // Deduplicates /page and /page/ as the same URL
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_trailing_slash_dedup";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'unique_urls.length' not available on result type
    }

    @Test
    void testCrawlUrlDeduplication() throws Exception {
        // Deduplicates URLs that differ only by fragment or query params
        var engine = Kreuzcrawl.createEngine(null);
        String url = System.getenv("MOCK_SERVER_URL") + "/fixtures/crawl_url_deduplication";
        var result = Kreuzcrawl.scrape(engine, url);
        // skipped: field 'pages.length' not available on result type
    }

}

File: e2e/elixir/test/crawl_test.exs
# E2e tests for category: crawl
defmodule E2e.CrawlTest do
  use ExUnit.Case, async: true

  describe "content_binary_skip" do
    test "Skips image and video content types gracefully" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'content.was_skipped' not available on result type
    end
  end

  describe "content_pdf_link_skip" do
    test "Encounters PDF link and skips or marks as document type" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'content.was_skipped' not available on result type
    end
  end

  describe "crawl_concurrent_depth" do
    test "Concurrent crawl respects max_depth limit" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
      # skipped: field 'stayed_on_domain' not available on result type
    end
  end

  describe "crawl_concurrent_limit" do
    test "Respects max concurrent requests limit during crawl" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_concurrent_max_pages" do
    test "Concurrent crawl respects max_pages budget" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_custom_headers" do
    test "Sends custom headers on all crawl requests" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_depth_one" do
    test "Follows links one level deep from start page" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
      # skipped: field 'stayed_on_domain' not available on result type
    end
  end

  describe "crawl_depth_priority" do
    test "Crawls in breadth-first order, processing depth-0 pages before depth-1" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_depth_two" do
    test "Crawls 3 levels deep (depth 0, 1, 2)" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_depth_two_chain" do
    test "Depth=2 crawl follows a chain of links across three levels" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_double_slash_normalization" do
    test "Normalizes double slashes in URL paths (//page to /page)" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'unique_urls.length' not available on result type
    end
  end

  describe "crawl_empty_page_no_links" do
    test "Crawl completes when child page has no outgoing links" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_exclude_path_pattern" do
    test "Skips URLs matching the exclude path pattern" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_external_links_ignored" do
    test "External links are discovered but not followed when stay_on_domain is true" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
      # skipped: field 'stayed_on_domain' not available on result type
    end
  end

  describe "crawl_fragment_stripping" do
    test "Strips #fragment from URLs for deduplication" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'unique_urls.length' not available on result type
    end
  end

  describe "crawl_include_path_pattern" do
    test "Only follows URLs matching the include path pattern" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_max_depth_zero" do
    test "max_depth=0 crawls only the seed page with no link following" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_max_pages" do
    test "Stops crawling at page budget limit" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_mixed_content_types" do
    test "Crawl handles links to non-HTML content types gracefully" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_multiple_redirects_in_traversal" do
    test "Multiple linked pages with redirects are handled during crawl traversal" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_query_param_dedup" do
    test "Deduplicates URLs with same query params in different order" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'unique_urls.length' not available on result type
    end
  end

  describe "crawl_redirect_in_traversal" do
    test "Links that redirect are followed during crawl traversal" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_self_link_no_loop" do
    test "Page linking to itself does not cause infinite crawl loop" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_single_page_no_links" do
    test "Crawling a page with no links returns only the seed page" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_stay_on_domain" do
    test "Does not follow external links when stay_on_domain is true" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
      # skipped: field 'stayed_on_domain' not available on result type
    end
  end

  describe "crawl_subdomain_exclusion" do
    test "Stays on exact domain and skips subdomain links" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
      # skipped: field 'stayed_on_domain' not available on result type
    end
  end

  describe "crawl_subdomain_inclusion" do
    test "Crawls subdomains when allow_subdomains is enabled" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end

  describe "crawl_trailing_slash_dedup" do
    test "Deduplicates /page and /page/ as the same URL" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'unique_urls.length' not available on result type
    end
  end

  describe "crawl_url_deduplication" do
    test "Deduplicates URLs that differ only by fragment or query params" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      # skipped: field 'pages.length' not available on result type
    end
  end
end
